### ISSUE TITLE
UDF Serializable error

### ISSUE THREAD

**Issue by yz-ark7** [scalapb/ScalaPB#841]
Hi,

I have encountered a serializable error when using the UDF function.

```scala
package ark7.etl.jobs

import com.ark7.data.user_behaviors.histories.sid_history.UserBehaviorSidHistory
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions.{col}
import scalapb.spark.ProtoSQL

import scalapb.spark.Implicits._
// import scalapb.spark.ProtoSQL.implicits._

object TestUDFJob {

  def main(args: Array[String]) = {
    val spark =
      SparkSession.builder().appName("test").master("local").getOrCreate()

    val u = ProtoSQL.udf { whatever: String => UserBehaviorSidHistory() }

    spark.read
      .parquet(
        "/Users/yz/Downloads/prod_client-user-behavior-event_daily_2020_05_30_part-00000-1440e467-8b6c-4ffa-938c-c8ef2b987740-c000.snappy.parquet",
      )
      .withColumn("test", u(col("sid")))
      .write
      .mode("overwrite")
      .save("/tmp/p5")
  }
}
```

```
20/06/03 08:56:16 INFO DAGScheduler: ResultStage 1 (save at test_udf.scala:26) failed in 0.057 s due to Job aborted due to stage failure: Task not serializable: java.io.NotSerializableException: scalapb.spark.Implicits$$anon$1
Serialization stack:
        - object not serializable (class: scalapb.spark.Implicits$$anon$1, value: scalapb.spark.Implicits$$anon$1@5a1765a9)
        - field (class: scalapb.spark.TypedEncoders$MessageTypedEncoder, name: $outer, type: interface scalapb.spark.TypedEncoders)
        - object (class scalapb.spark.TypedEncoders$MessageTypedEncoder, scalapb.spark.TypedEncoders$MessageTypedEncoder@72b7516a)
        - field (class: frameless.functions.FramelessUdf, name: rencoder, type: class frameless.TypedEncoder)
        - object (class frameless.functions.FramelessUdf, FramelessUdf(input[16, string, true].toString))
        - element of array (index: 1)
        - array (class [Ljava.lang.Object;, size 23)
        - element of array (index: 1)
        - array (class [Ljava.lang.Object;, size 3)
        - field (class: java.lang.invoke.SerializedLambda, name: capturedArgs, type: class [Ljava.lang.Object;)
        - object (class java.lang.invoke.SerializedLambda, SerializedLambda[capturingClass=class org.apache.spark.sql.execution.WholeStageCodegenExec, functionalInterfaceMethod=scala/Function2.apply:(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;, implementation=invokeStatic org/apache/spark/sql/execution/Wh
oleStageCodegenExec.$anonfun$doExecute$4$adapted:(Lorg/apache/spark/sql/catalyst/expressions/codegen/CodeAndComment;[Ljava/lang/Object;Lorg/apache/spark/sql/execution/metric/SQLMetric;Ljava/lang/Object;Lscala/collection/Iterator;)Lscala/collection/Iterator;, instantiatedMethodType=(Ljava/lang/Object;Lscala/collection
/Iterator;)Lscala/collection/Iterator;, numCaptured=3])
        - writeReplace data (class: java.lang.invoke.SerializedLambda)
        - object (class org.apache.spark.sql.execution.WholeStageCodegenExec$$Lambda$7610/0x0000000802f9a840, org.apache.spark.sql.execution.WholeStageCodegenExec$$Lambda$7610/0x0000000802f9a840@152e1538)
        - element of array (index: 0)
        - array (class [Ljava.lang.Object;, size 1)
        - field (class: java.lang.invoke.SerializedLambda, name: capturedArgs, type: class [Ljava.lang.Object;)
        - object (class java.lang.invoke.SerializedLambda, SerializedLambda[capturingClass=class org.apache.spark.rdd.RDD, functionalInterfaceMethod=scala/Function3.apply:(Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;, implementation=invokeStatic org/apache/spark/rdd/RDD.$anonfun$mapPartit
ionsWithIndex$2$adapted:(Lscala/Function2;Lorg/apache/spark/TaskContext;Ljava/lang/Object;Lscala/collection/Iterator;)Lscala/collection/Iterator;, instantiatedMethodType=(Lorg/apache/spark/TaskContext;Ljava/lang/Object;Lscala/collection/Iterator;)Lscala/collection/Iterator;, numCaptured=1])
        - writeReplace data (class: java.lang.invoke.SerializedLambda)
        - object (class org.apache.spark.rdd.RDD$$Lambda$7612/0x0000000802f9b040, org.apache.spark.rdd.RDD$$Lambda$7612/0x0000000802f9b040@449c1319)
        - field (class: org.apache.spark.rdd.MapPartitionsRDD, name: f, type: interface scala.Function3)
        - object (class org.apache.spark.rdd.MapPartitionsRDD, MapPartitionsRDD[3] at save at test_udf.scala:26)
        - field (class: scala.Tuple2, name: _1, type: class java.lang.Object)
        - object (class scala.Tuple2, (MapPartitionsRDD[3] at save at test_udf.scala:26,org.apache.spark.sql.execution.datasources.FileFormatWriter$$$Lambda$7613/0x0000000802f9b840@137a592f))
```

If I switch to use ProtoSQL.implicits._

```scala
// import scalapb.spark.Implicits._
import scalapb.spark.ProtoSQL.implicits._
```

```
20/06/03 08:57:27 INFO DAGScheduler: ResultStage 1 (save at test_udf.scala:26) failed in 0.058 s due to Job aborted due to stage failure: Task not serializable: java.io.NotSerializableException: scalapb.spark.ProtoSQL$$anon$1$$anon$2
Serialization stack:
        - object not serializable (class: scalapb.spark.ProtoSQL$$anon$1$$anon$2, value: scalapb.spark.ProtoSQL$$anon$1$$anon$2@f2bd85c)
        - field (class: scalapb.spark.TypedEncoders$MessageTypedEncoder, name: $outer, type: interface scalapb.spark.TypedEncoders)
        - object (class scalapb.spark.TypedEncoders$MessageTypedEncoder, scalapb.spark.TypedEncoders$MessageTypedEncoder@6ee241a4)
        - field (class: frameless.functions.FramelessUdf, name: rencoder, type: class frameless.TypedEncoder)
        - object (class frameless.functions.FramelessUdf, FramelessUdf(input[16, string, true].toString))
        - element of array (index: 1)
        - array (class [Ljava.lang.Object;, size 23)
        - element of array (index: 1)
        - array (class [Ljava.lang.Object;, size 3)
        - field (class: java.lang.invoke.SerializedLambda, name: capturedArgs, type: class [Ljava.lang.Object;)
        - object (class java.lang.invoke.SerializedLambda, SerializedLambda[capturingClass=class org.apache.spark.sql.execution.WholeStageCodegenExec, functionalInterfaceMethod=scala/Function2.apply:(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;, implementation=invokeStatic org/apache/spark/sql/execution/Wh
oleStageCodegenExec.$anonfun$doExecute$4$adapted:(Lorg/apache/spark/sql/catalyst/expressions/codegen/CodeAndComment;[Ljava/lang/Object;Lorg/apache/spark/sql/execution/metric/SQLMetric;Ljava/lang/Object;Lscala/collection/Iterator;)Lscala/collection/Iterator;, instantiatedMethodType=(Ljava/lang/Object;Lscala/collection
/Iterator;)Lscala/collection/Iterator;, numCaptured=3])
        - writeReplace data (class: java.lang.invoke.SerializedLambda)
        - object (class org.apache.spark.sql.execution.WholeStageCodegenExec$$Lambda$7703/0x0000000802f90840, org.apache.spark.sql.execution.WholeStageCodegenExec$$Lambda$7703/0x0000000802f90840@6be51ac4)
        - element of array (index: 0)
        - array (class [Ljava.lang.Object;, size 1)
        - field (class: java.lang.invoke.SerializedLambda, name: capturedArgs, type: class [Ljava.lang.Object;)
        - object (class java.lang.invoke.SerializedLambda, SerializedLambda[capturingClass=class org.apache.spark.rdd.RDD, functionalInterfaceMethod=scala/Function3.apply:(Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;, implementation=invokeStatic org/apache/spark/rdd/RDD.$anonfun$mapPartit
ionsWithIndex$2$adapted:(Lscala/Function2;Lorg/apache/spark/TaskContext;Ljava/lang/Object;Lscala/collection/Iterator;)Lscala/collection/Iterator;, instantiatedMethodType=(Lorg/apache/spark/TaskContext;Ljava/lang/Object;Lscala/collection/Iterator;)Lscala/collection/Iterator;, numCaptured=1])
        - writeReplace data (class: java.lang.invoke.SerializedLambda)
        - object (class org.apache.spark.rdd.RDD$$Lambda$7705/0x0000000802fa0040, org.apache.spark.rdd.RDD$$Lambda$7705/0x0000000802fa0040@79c18b6f)
        - field (class: org.apache.spark.rdd.MapPartitionsRDD, name: f, type: interface scala.Function3)
        - object (class org.apache.spark.rdd.MapPartitionsRDD, MapPartitionsRDD[3] at save at test_udf.scala:26)
        - field (class: scala.Tuple2, name: _1, type: class java.lang.Object)
        - object (class scala.Tuple2, (MapPartitionsRDD[3] at save at test_udf.scala:26,org.apache.spark.sql.execution.datasources.FileFormatWriter$$$Lambda$7706/0x0000000802fa0840@75aff4d9))
```

Did I write something wrong? Thanks!

**Comment by thesamet** [comment]
Thanks for the report. Can you help to reproduce this issue by cloning [scalapb/sparksql-scalapb](https://github.com/scalapb/sparksql-scalapb), and adding a failing unit test there: https://github.com/scalapb/sparksql-scalapb/blob/66f87c4a0557b396160f8eee206b13b61f68a9f6/sparksql-scalapb/src/test/scala/PersonSpec.scala#L280-L296 . The link shows a similar UDF test that is currently passing. Perhaps this would help you in debugging this. Then you can run the test suite locally in sbt.

**Comment by yz-ark7** [comment]
Thanks @thesamet for the example. This is very interesting. After played with it a bit, I found

The following code works:

```scala
    val u = ProtoSQL.udf { whatever: String => UserBehaviorSidHistory() }

    val df = Seq(("hello"), ("world"))
      .toDF("h")
      .withColumn("foo", u(col("h")))
      .write
      .mode("overwrite")
      .save("/tmp/p5")

```

While below doesn't:
```scala
    val u = ProtoSQL.udf { whatever: String => UserBehaviorSidHistory() }

    val df =
      spark.read
        .parquet(
          "/Users/yz/Downloads/prod_client-user-behavior-event_daily_2020_05_30_part-00000-1440e467-8b6c-4ffa-938c-c8ef2b987740-c000.snappy.parquet",
        )
        .withColumn("h", lit("1"))
        .select("h")
        .withColumn("foo", u(col("h")))
        .write
        .mode("overwrite")
        .save("/tmp/p5")
```

This is very strange, looks like a conflict with `parquet`?

```
Serialization stack:
        - object not serializable (class: scalapb.spark.ProtoSQL$$anon$1$$anon$2, value: scalapb.spark.ProtoSQL$$anon$1$$anon$2@465fba5a)
        - field (class: scalapb.spark.TypedEncoders$MessageTypedEncoder, name: $outer, type: interface scalapb.spark.TypedEncoders)
        - object (class scalapb.spark.TypedEncoders$MessageTypedEncoder, scalapb.spark.TypedEncoders$MessageTypedEncoder@580d5e5b)
        - field (class: frameless.functions.FramelessUdf, name: rencoder, type: class frameless.TypedEncoder)
        - object (class frameless.functions.FramelessUdf, FramelessUdf(1.toString))
        - element of array (index: 3)
        - array (class [Ljava.lang.Object;, size 28)
        - element of array (index: 1)
        - array (class [Ljava.lang.Object;, size 3)
        - field (class: java.lang.invoke.SerializedLambda, name: capturedArgs, type: class [Ljava.lang.Object;)
```

**Comment by yz-ark7** [comment]
Also, tried json input, still not working.

```scala
    val df =
      spark.read
        .json(
          "/Users/yz/Downloads/prod_allosaurus-access-log_daily_2020_05_15_part-00003-49cc4226-1497-48ce-9338-c6d15db03579-c000.json",
        )
        .withColumn("h", lit("1"))
        .select("h")
        .withColumn("foo", u(col("h")))
        .write
        .mode("overwrite")
        .save("/tmp/p5")
```

```
20/06/03 10:28:17 INFO DAGScheduler: ResultStage 1 (save at test_udf.scala:40) failed in 0.026 s due to Job aborted due to stage failure: Task not serializable: java.io.NotSerializableException: scalapb.spark.ProtoSQL$$anon$1$$anon$2
Serialization stack:
        - object not serializable (class: scalapb.spark.ProtoSQL$$anon$1$$anon$2, value: scalapb.spark.ProtoSQL$$anon$1$$anon$2@7cc43fae)
        - field (class: scalapb.spark.TypedEncoders$MessageTypedEncoder, name: $outer, type: interface scalapb.spark.TypedEncoders)
        - object (class scalapb.spark.TypedEncoders$MessageTypedEncoder, scalapb.spark.TypedEncoders$MessageTypedEncoder@63a07871)
        - field (class: frameless.functions.FramelessUdf, name: rencoder, type: class frameless.TypedEncoder)
        - object (class frameless.functions.FramelessUdf, FramelessUdf(1.toString))
        - element of array (index: 2)
        - array (class [Ljava.lang.Object;, size 27)
        - element of array (index: 1)
        - array (class [Ljava.lang.Object;, size 3)
        - field (class: java.lang.invoke.SerializedLambda, name: capturedArgs, type: class [Ljava.lang.Object;)
```

**Comment by thesamet** [comment]
Very Interesting. Can you send a PR with a new unit test that breaks?

**Comment by andyzhau** [comment]
@thesamet added: https://github.com/scalapb/sparksql-scalapb/pull/94

**Comment by thesamet** [comment]
This has been fixed in sparksql-scalapb 0.10.3.
